### DISCUSSION TITLE
Leonardo/ProMicro 32U4 issue with DualMerger example

### DISCUSSION THREAD

**Discussion by Petrus125** [FortySevenEffects/arduino_midi_library#306] [Q&A] [answered]
Hello, I'm currently developing over the DualMerger.ino example. Works perfect with UNO Board, but not working under Sparkfun Pro Micro. Guess has to do with interrupts for Serial input. ProMicro needs any special request? is  it compatible with this lib?
greets

```c++
#include <MIDI.h>
#include "Controller.h"
#include <SoftwareSerial.h>

    SoftwareSerial softSerial(2,3); // Software Serial Port Pins (2,3) (Rx,Tx)
    MIDI_CREATE_INSTANCE(HardwareSerial, Serial,     midiA);
    MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial, midiB);
    MIDI_CREATE_DEFAULT_INSTANCE();   

void setup()  {

    attachInterrupt(digitalPinToInterrupt(2), loop, CHANGE);  
    
    // Initiate MIDI communications, listen to all channels
    midiA.begin(MIDI_CHANNEL_OMNI);
    midiB.begin(MIDI_CHANNEL_OMNI);
    MIDI.begin (MIDI_CHANNEL_OFF);
    
  }

void loop()

{
    if (midiA.read())
    {
        // Thru on A has already pushed the input message to out A.
        // Forward the message to out B as well.
        midiB.send(midiA.getType(),
                   midiA.getData1(),
                   midiA.getData2(),
                   midiA.getChannel());
    }

    if (midiB.read())
    {
        // Thru on B has already pushed the input message to out B.
        // Forward the message to out A as well.
        midiA.send(midiB.getType(),
                   midiB.getData1(),
                   midiB.getData2(),
                   midiB.getChannel());
    }

  }
```

**Comment by lathoub** [comment]
Hello, I'm not familiar with the Sparkfun Pro Micro.

But looking at the code, you instantiate 3 MIDI ports: midiA, midiB and MIDI. Yet you only use 2 in your code: midiA and midiB.

I also have a suspicion that midiA (using hardware serial) overlaps with MIDI (also using hardware serial, yet hidden by the MACRO `MIDI_CREATE_DEFAULT_INSTANCE();`) - even when not listening (`MIDI_CHANNEL_OFF`, the code still runs)

So: delete `MIDI_CREATE_DEFAULT_INSTANCE();`  and `MIDI.begin (MIDI_CHANNEL_OFF);`

Note: i can't see in your code why you need the interrupt.

**Reply by Petrus125** [reply]
Thank you lathoub, maybe, I will check again with the real midi message sent with the hardware. The point is that I still don't see why same code runs perfect with an UNO R3 bd genuine and not with the ProMicro. The only difference is the board.And some forums point to need of extra interrupt comands needed. I still don't see why. All 32U4 handle with pins 0,1,2,3,7.
Thanks again,

**Comment by lathoub** [comment]
Do you have another ProMicro to test this on?

**Reply by Petrus125** [reply]
Unfortunately not. I've ordered a couple more, with 32U4, 328P and CH340, just to discard HW.  Let you know when I get those and try. But all test lead to ProMicro somehow, sure- there must be a workaround.

**Comment by franky47** [comment]
Just to rule some things out, could you test the following things?

- Just writing on SoftwareSerial
- Just reading on SoftwareSerial (and something like blink an LED on received messages)
- Just writing on HardwareSerial
- Just reading on HardwareSerial

**Reply by franky47** [reply]
I'm not sure I'm following your setup. If you don't need SoftwareSerial write, and don't need HardwareSerial read, why have two instances? Also, you mention USB MIDI, but neither the Uno nor the Pro Micro have USB MIDI capabilities.

Could you use the HardwareSerial (more reliable) for both reading CC & Notes, then forward them to the output (either using the SoftThru feature or by re-forwarding selected message types, where the upcoming [map/filter API](https://github.com/FortySevenEffects/arduino_midi_library/issues/40#issuecomment-647012093) would help)? You would then also read switches and write messages to the HardwareSerial TX. The library will ensure messages are not mangled.

**Reply by franky47** [reply]
The Pro Micro (boasting an ATmega 32U4) definitely has native USB MIDI capabilities, sorry about that.

Though it will need a dedicated interface to be used as such, see: https://github.com/lathoub/Arduino-USBMIDI

In any way, it will act as a USB MIDI device, not a host (ie: you won't be able to plug a USB MIDI controller on it and talk to it).

**Reply by Petrus125** [reply]
Hi Franky,  yes ATmega 32U4 has native USB MIDI, sure, full midi ClassCompliant, no drivers needed. Even The Ipad works over it. 
And the UNO R3, that Im using for HW testing, mod with the Pimp-Mocolufa https://github.com/kuwatay/mocolufaFW update, also turns CCompliant.
For the usb Host port, I'm using an external USB-to-serial converter chip. Works perfect. This makes programing more simple to me, just a musician, no high skill programer, so some of your helping choices are bit dificult for me.
With the latest provided sketch, just would need the ProMicro to work like the UNO......and voilà!

**Comment by Petrus125** [comment]
More R&D..... a genuine Leonardo with a 32U4 also doesn't work.... same as ProMicro 32U4 issue. 
Only the UNO does the job. Have no Idea, but seems all 32U4. Related to library code or compatibility?

**Comment by franky47** [comment]
I didn't really understand your response concerning USB MIDI, but if you want to use the 32U4 to act as a USB MIDI device and talk with a USB host, you'll really have to use the USBMIDI library made by @lathoub in addition to this one, check [the examples](https://github.com/lathoub/Arduino-USBMIDI/blob/master/examples/MIDI_DIN2USB/MIDI_DIN2USB.ino) there.

It has worked on the Uno as it talks using HardwareSerial to the USB transceiver chip that deals with USB device enumeration (as a MIDI device in your case as you patched it to behave like so). But this transceiver bridge does not exist on the 32U4, which has a built-in USB interface.

**Reply by Petrus125** [reply]
Hello, yes, probably problem is here. But in order to run the shown sketch MIDI.h lib is needed. But cannot use MIDI.h and USB-Midi.h both at same time; compile error. If you please just show me the way to make the upper sketch run under  any 32U4, (leonardo, micro or promicro) the I get the rest. Sure!

**Reply by franky47** [reply]
`#include <USB-MIDI.h>` already includes `<MIDI.h>` internally, no need to re-include it. Make sure to follow the [installation instructions](https://github.com/lathoub/Arduino-USBMIDI) as it's a different library that wraps the MIDI library.

**Reply by Petrus125** [reply]
You're absolutely right. But the most important is to remove the MIDI.h folder from the library folder; Or compilation error come up if both libs are coexist.

**Reply by Petrus125** [reply]
There is a diference between UNO and Leonardo: Leonardo has 2 HW serial ports, Uno one. So to refer to the HW port, leonardo refers Serial1. I already md the code accordingly, but stil not working. And here, I'm bit confused. The sketch is as follows:

// This example shows how to create two instances of the library to create a merger.
// There are two MIDI couples of IO, A and B, each using thru and merging with the
// input from the other node. The result is the following:
// A out = A in + B in
// B out = B in + A in
   
    #include <USB-MIDI.h>    
    #include <SoftwareSerial.h>
    SoftwareSerial softSerial(2,3); // Software Serial Port Pins (2,3) (Rx,Tx)
    MIDI_CREATE_INSTANCE(HardwareSerial, Serial1,    midiA); //Atmega32U4 uses Serial1 . Uno Serial.
    MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial, midiB);

void setup(){
    // Initiate MIDI communications, listen to all channels
    midiA.begin(MIDI_CHANNEL_OMNI);
    midiB.begin(MIDI_CHANNEL_OMNI);
}

void loop(){
    if (midiA.read()) {
        // Thru on A has already pushed the input message to out A.
        // Forward the message to out B as well.
        midiB.send(midiA.getType(),
                   midiA.getData1(),
                   midiA.getData2(),
                   midiA.getChannel());
    }

    if (midiB.read())    {
        // Thru on B has already pushed the input message to out B.
        // Forward the message to out A as well.
        midiA.send(midiB.getType(),
                   midiB.getData1(),
                   midiB.getData2(),
                   midiB.getChannel());
    }
}

**Comment by Petrus125** [comment, marked as answer]
Finally got it! well almost there. The clue is a limitation of the SoftwareSerial library:
"Not all pins on the Leonardo and Micro boards support change interrupts, so only the following can be used for RX: 8, 9, 10, 11, 14
(MISO), 15 (SCK), 16 (MOSI)."
So using pin 8, instead of pin 2, works fine.

**Comment by Petrus125** [comment]
And finally, still need how to add midi over usb (merged with midiA+midiB) both to Serial1. Can you help here?

```c++
#include <USB-MIDI.h>    
#include <SoftwareSerial.h>
SoftwareSerial softSerial(8,9); // Software Serial Port Pins (8,9) (Rx,Tx)
MIDI_CREATE_INSTANCE(HardwareSerial, Serial1,    midiA); //Atmega32U4 uses Serial1 . Uno Serial.
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial, midiB);

void setup(){
// Initiate MIDI communications, listen to all channels
midiA.begin(MIDI_CHANNEL_OMNI);
midiB.begin(MIDI_CHANNEL_OMNI);
}

void loop(){
if (midiA.read()) {
// Thru on A has already pushed the input message to out A.
// Forward the message to out B as well.
midiB.send(midiA.getType(),
midiA.getData1(),
midiA.getData2(),
midiA.getChannel());
}

if (midiB.read())    {
    // Thru on B has already pushed the input message to out B.
    // Forward the message to out A as well.
    midiA.send(midiB.getType(),
               midiB.getData1(),
               midiB.getData2(),
               midiB.getChannel());
}

}
```

**Comment by lathoub** [comment]
@Petrus125 apart from the USB-MIDI include, I don't see any reference to a USB MIDI instance as described on https://github.com/lathoub/Arduino-USBMIDI. The above code only references serial MIDI.

Have you looked at the example franky47 pointed you to ([example](https://github.com/lathoub/Arduino-USBMIDI/blob/master/examples/MIDI_DIN2USB/MIDI_DIN2USB.ino))

**Reply by Petrus125** [reply]
Yes, I've looked this example. I tested it and the results are:
- Does DIN to USB perfect
- Does Din to Din perfect.
- Does not USB to Din.  This is what I need at this moment. And from the code, it should.
did you test? what mod need to do USB toDIN?

**Reply by vamix2012** [reply]
Because someone said that the 32u4 does not have the usb host capability, so you can't talk to it....
You need something like SAMD architecture  microcontroller(NANO 33 or XIAO Seed) they work well for this kind of tasks as long as I know because they can take usb midi input and route it to the DIN MIDI....

**Reply by vamix2012** [reply]
Check this library 
https://github.com/marchingband/USB_Host_Library_SAMD/tree/master/examples/USBH_MIDI

**Reply by Petrus125** [reply]
Thank you Vamix. Finally redesigned all from scratch using SurfaceControl lib examples. Finally all midi merge flows up and running fine for Leonardo, ProMicro, all. Final project at FB Boss RC-600 group.
